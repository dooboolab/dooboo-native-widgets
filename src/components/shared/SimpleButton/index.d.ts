import { ThemeProps } from 'styled-components';

interface TextThemeType {
  fontColor: string;
}

interface ButtonThemeType {
  backgroundColor: string;
  borderColor: string;
}

interface ButtonType extends ThemeProps<ButtonThemeType> {
  theme: ButtonThemeType;
}

interface TextType extends ThemeProps<TextThemeType>{
  theme: TextThemeType;
}

interface ThemeType extends ButtonThemeType, TextThemeType {
  [key: string]: any;
}

interface StatefulThemeType extends ThemeType {
  INVERTED: ThemeType;
  DISABLED: ThemeType;
}
