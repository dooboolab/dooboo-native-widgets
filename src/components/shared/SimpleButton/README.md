# SimpleButton

> [SimpleButton] component that can be used inside product. Has basic features like `loading` state, `disabled` state and also has ability to put `img` to left-hand which is used very often.

## Props

|                | necessary | types               | default |
| -------------- | --------- | ------------------- | ------- |
| testID         |           | string              |         |
| themeType      |           | string              |         |
| width          |           | number              |         |
| height         |           | number              |         |
| inverted       |           | boolean             |         |
| isLoading      |           | boolean             |         |
| isDisabled     |           | boolean             |         |
| iconSrc        |           | ImageSourcePropType |         |
| indicatorColor |           | string              |         |
| activeOpacity  |           | number              |         |
| children       |           | string              |         |
| onClick        |           | func                |         |

## Getting started

- Import

  ```javascript
  import { SimpleButton } from '@dooboo-ui/native';
  ```

- Usage
  ```javascript
  function Page(props: Props) {
    return (
      <Container>
        <SimpleButton
          testID='btn'
          isLoading={false}
          onPress={() => {}}
        >
          😀 😎 👍 💯
        </SimpleButton>
        <SimpleButton
          style={{
            marginVertical: 40,
          }}
          isDisabled={true}
          onPress={() => {}}
        >
          This is disabled!!
        </SimpleButton>
        <Button
          testID='btnGoogle'
          iconSrc={IC_GOOGLE}
          isLoading={googleLoading}
          indicatorColor='#023059'
          onPress={() => {
            setGoogleLoading(true);
            const timeout = setTimeout(() => {
              setGoogleLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
        >
          GOOGLE SIGN IN
        </SimpleButton>
        <SimpleButton
          testID='btnFacebook'
          icon={IC_FACEBOOK}
          indicatorColor='#023059'
          isLoading={facebookLoading}
          style={{
            marginTop: 40,
            backgroundColor: '#ccc',
            borderWidth: 0.5,
            borderRadius: 0,
          }}
          onPress={() => {
            setFacebookLoading(true);
            const timeout = setTimeout(() => {
              setFacebookLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
        >
          FACEBOOK SIGN IN
        </SimpleButton>
      </Container>
    );
  }
  ```

```

```
