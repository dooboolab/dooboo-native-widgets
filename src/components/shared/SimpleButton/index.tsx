import {
  ActivityIndicator,
  ImageSourcePropType,
  TouchableOpacity,
} from 'react-native';
import styled, { ThemeProps } from 'styled-components/native';

import React from 'react';

const COLOR = {
  WHITE: 'white',
  DODGERBLUE: 'rgb(58, 139, 255)',
};

export const THEME_TYPE = {
  LIGHT: 'light',
  DARK: 'dark',
};

export const THEME = {
  [THEME_TYPE.LIGHT]: {
    backgroundColor: COLOR.DODGERBLUE,
    borderColor: COLOR.WHITE,
    fontColor: COLOR.WHITE,
    INVERTED: {
      backgroundColor: COLOR.WHITE,
      borderColor: COLOR.DODGERBLUE,
      fontColor: COLOR.DODGERBLUE,
    },
  },
  [THEME_TYPE.DARK]: {
    backgroundColor: COLOR.DODGERBLUE,
    borderColor: COLOR.WHITE,
    fontColor: COLOR.WHITE,
    INVERTED: {
      backgroundColor: COLOR.WHITE,
      borderColor: COLOR.DODGERBLUE,
      fontColor: COLOR.DODGERBLUE,
    },
  },
};

interface Props {
  testID: string;
  themeType: string;
  width: number;
  height: number;
  inverted?: boolean;
  isLoading: boolean;
  isDisabled: boolean;
  iconSrc?: ImageSourcePropType;
  indicatorColor: string;
  activeOpacity: number;
  children: string;
  onPress: () => void;
}

interface TextTheme {
  fontColor: string;
}
interface Text extends ThemeProps<TextTheme>{
  theme: TextTheme;
}

interface ButtonTheme {
  backgroundColor: string;
  borderColor: string;
}
interface Button extends ThemeProps<ButtonTheme> {
  width?: number;
  height?: number;
}

const StyledButton = styled.View<Button>`
  width: ${({ width }): string => `${width}px`};
  height: ${({ height }): string => `${height}px`};
  background-color: ${({ theme }): string => theme.backgroundColor};
  border-color: ${({ theme }): string => theme.borderColor};
  border-radius: 4px;
  border-width: 1px;
  justify-content: center;
  align-items: center;
`;

const StyledButtonDisabled = styled.View<Button>`
  width: ${({ width }): string => `${width}px`};
  height: ${({ height }): string => `${height}px`};
  background-color: rgb(243, 243, 243);
  align-self: center;
  border-radius: 4px;
  border-width: 2px;
  border-color: #333;
  align-items: center;
  justify-content: center;
`;

const Text = styled.Text<Text>`
  font-size: 14px;
  font-weight: bold;
  color: ${({ theme }): string => theme.fontColor};
`;

const Icon = styled.Image`
  width: 24px;
  height: 24px;
  position: absolute;
  left: 16px;
`;

function SimpleButton(props: Props): React.ReactElement {
  const {
    testID,
    isDisabled,
    inverted,
    indicatorColor,
    iconSrc,
    activeOpacity,
    onPress,
    isLoading,
    width,
    height,
    children: text,
    themeType,
  } = props;

  const theme = inverted ? THEME[themeType].INVERTED : THEME[themeType];

  if (isDisabled) {
    return (
      <StyledButtonDisabled
        testID={testID}
        theme={theme}
        width={width}
        height={height}
      >
        {<Text theme={theme}>{text}</Text>}
      </StyledButtonDisabled>
    );
  }

  return (
    <TouchableOpacity
      testID={testID}
      activeOpacity={activeOpacity}
      onPress={onPress}
    >
      <StyledButton
        theme={theme}
        width={width}
        height={height}
      >
        {iconSrc && <Icon source={iconSrc} />}
        {!isLoading && <Text theme={theme}>{text}</Text>}
        {isLoading && <ActivityIndicator size='small' color={indicatorColor} />}
      </StyledButton>
    </TouchableOpacity>
  );
}

SimpleButton.defaultProps = {
  themeType: THEME_TYPE.LIGHT,
  width: 136,
  height: 60,
  isLoading: false,
  isDisabled: false,
  indicatorColor: 'white',
  activeOpacity: 0.5,
};

export default SimpleButton;
