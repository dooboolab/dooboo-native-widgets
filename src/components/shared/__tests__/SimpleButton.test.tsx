import 'react-native';

import * as React from 'react';

import {
  RenderResult,
  act,
  fireEvent,
  render,
} from '@testing-library/react-native';
import SimpleButton, { THEME, THEME_TYPE } from '../SimpleButton';

import { ThemeProvider } from 'styled-components/native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

describe('[SimpleButton]', () => {
  let cnt = 1;

  beforeEach(() => {
    props = {
      onPress: (): number => cnt++,
      testID: 'btn',
    };
    component = (
      <ThemeProvider theme={THEME[THEME_TYPE.LIGHT]}>
        <SimpleButton {...props} />
      </ThemeProvider>
    );
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    beforeEach(() => {
      testingLib = render(component);
    });

    it('should simulate onPress', () => {
      const btn = testingLib.queryByTestId('btn');
      act(() => {
        fireEvent.press(btn);
        fireEvent.press(btn);
      });
      expect(cnt).toBe(3);
    });
  });
});
