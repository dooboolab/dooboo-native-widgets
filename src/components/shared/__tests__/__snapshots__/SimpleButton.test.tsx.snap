// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SimpleButton] renders without crashing 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    height={60}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(58, 139, 255)",
          "borderColor": "white",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 60,
          "justifyContent": "center",
          "width": 136,
        },
      ]
    }
    theme={
      Object {
        "INVERTED": Object {
          "backgroundColor": "white",
          "borderColor": "rgb(58, 139, 255)",
          "fontColor": "rgb(58, 139, 255)",
        },
        "backgroundColor": "rgb(58, 139, 255)",
        "borderColor": "white",
        "fontColor": "white",
      }
    }
    width={136}
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "bold",
          },
        ]
      }
      theme={
        Object {
          "INVERTED": Object {
            "backgroundColor": "white",
            "borderColor": "rgb(58, 139, 255)",
            "fontColor": "rgb(58, 139, 255)",
          },
          "backgroundColor": "rgb(58, 139, 255)",
          "borderColor": "white",
          "fontColor": "white",
        }
      }
    />
  </View>
</View>
`;
